File: earthquake.py
import streamlit as st
import numpy as np
import matplotlib.pyplot as plt
import time

def page_earthquake():
    st.header("🏗️ Simulasi Gempa Gedung (SNI 1726)")
    st.write("""
    **Analisis Statik Ekuivalen Berdasarkan SNI 1726:2019**  
    Simulasi ini menghitung Gaya Geser Dasar Seismik ($V$) dan Simpangan Atap Maksimal (Roof Drift) menggunakan parameter respons spektrum, Sistem Penahan Gaya Seismik ($R$), dan mengecek batasan izin simpangan bangunan.
    """)
    
    col1, col2 = st.columns([1, 2])
    
    with col1:
        st.subheader("Parameter Gedung & Gempa")
        tinggi = st.slider("Jumlah Lantai Gedung", 3, 20, 10, key="eq_tinggi", help="Tinggi asumsi per lantai adalah 4 meter.")
        material = st.selectbox("Sistem Penahan Gaya Seismik", [
            "Rangka Beton Pemikul Momen Khusus (SRPMK, R=8)",
            "Rangka Beton Pemikul Momen Menengah (SRPMM, R=5)",
            "Rangka Beton Pemikul Momen Biasa (SRPMB, R=3)" 
        ])
        
        magnitudo_slider = st.slider("Kekuatan Gempa (Setara Percepatan SDS)", 1.0, 10.0, 6.0, step=0.5, help="Angka simulasi ini akan dipetakan ke nilai Spektrum Respons Desain (SDS).")
        
        # Mapping properties
        if "SRPMK" in material:
            R = 8.0
            Cd = 5.5
            damping = 0.05
        elif "SRPMM" in material:
            R = 5.0
            Cd = 4.5
            damping = 0.05
        else:
            R = 3.0
            Cd = 2.5
            damping = 0.05
            
        Sds = (magnitudo_slider / 10.0) * 1.5 # 0.15g to 1.5g
        Ie = 1.0 # Faktor Keutamaan
        
        st.info(f"""
        💡 **Parameter Analisis:**
        * **SDS** = {Sds:.2f} g
        * **Faktor Modifikasi (R)** = {R}
        * **Faktor Pembesaran (Cd)** = {Cd}
        """)
        
    with col2:
        plot_placeholder = st.empty()
        status_placeholder = st.empty()
        
        # SNI 1726 Calculations
        floor_height = 4.0
        hn = tinggi * floor_height
        W = tinggi * 2500.0 # kN
        
        Cs = Sds / (R / Ie)
        if Cs < 0.044 * Sds * Ie:
            Cs = 0.044 * Sds * Ie
            
        V_base = Cs * W
        
        # Empirical approximation for demonstration
        Delta_elastic = (Cs * hn) / 20.0 
        max_drift = (Cd * Delta_elastic) / Ie
        allowable_drift = 0.020 * hn
        
        # For animation physics
        T_period = 0.0466 * (hn ** 0.9)
        freq = 1.0 / T_period
        
        def plot_building(amplitude_t, t_val=0):
            fig, ax = plt.subplots(figsize=(5, 6))
            ax.axhline(0, color='black', lw=3)
            
            y_coords = np.linspace(0, hn, tinggi + 1)
            # Mode shape approximation: y^1.5
            x_coords = amplitude_t * ((y_coords / hn) ** 1.5)
            
            offset = 2.0
            # Draw columns and floor slabs
            ax.plot(x_coords - offset, y_coords, color='#0068C9', lw=4, marker='o')
            ax.plot(x_coords + offset, y_coords, color='#0068C9', lw=4, marker='o')
            for y_val, x_val in zip(y_coords, x_coords):
                ax.plot([x_val - offset, x_val + offset], [y_val, y_val], color='gray', linestyle='-', alpha=0.8, lw=3)
            
            ax.set_xlim(-10, 10)
            ax.set_ylim(-1, hn + 5)
            ax.set_xlabel("Simpangan Atap (Meter)")
            ax.set_ylabel("Tinggi Gedung (Meter)")
            ax.set_title(f"Goyangan Gempa | t = {t_val:.2f}s | Base Shear = {V_base:.0f} kN")
            ax.grid(True, linestyle=':', alpha=0.6)
            return fig
            
        # Draw initial static state
        fig_init = plot_building(0.0, 0.0)
        plot_placeholder.pyplot(fig_init)
        plt.close(fig_init)
        
        status_placeholder.info(f"👉 Klik tombol untuk melihat respon struktur. Batas izin simpangan SNI: **{allowable_drift:.2f} meter**.")
        
        if st.button("🚀 MULAI SIMULASI GEMPA (SNI 1726)", use_container_width=True):
            t_steps = np.linspace(0, 8, 40)
            
            for t in t_steps:
                # Envelope: damped harmonic
                amplitude = max_drift * np.exp(-damping * 2.0 * t) * np.sin(2 * np.pi * freq * t)
                
                fig = plot_building(amplitude, t)
                plot_placeholder.pyplot(fig)
                plt.close(fig)
                
                # Dynamic feedback
                if max_drift > allowable_drift * 1.5:
                    status_placeholder.error(f"🚨 BAHAYA RUNTUH: Simpangan Atap ({max_drift:.2f}m) sangat melebihi batas izin ({allowable_drift:.2f}m). Sistem {material.split('(')[1].split(',')[0]} GAGAL MENAHAN GEMPA!")
                    break
                elif max_drift > allowable_drift:
                    status_placeholder.warning(f"⚠️ TIDAK MEMENUHI SNI: Simpangan Atap ({max_drift:.2f}m) melampaui batas izin ({allowable_drift:.2f}m). Kolom retak parah.")
                else:
                    status_placeholder.success(f"✅ MEMENUHI SNI 1726: Simpangan Atap ({max_drift:.2f}m) AMAN di bawah batas izin ({allowable_drift:.2f}m). Struktur Daktail!")
                time.sleep(0.05)

File: civil-scripts-web/src/lib/physics/earthquake.ts
/**
 * Earthquake physics engine — Real Case (SNI 1726:2019)
 */

export interface EarthquakeParams {
  floors: number
  material: "SRPMK" | "SRPMM" | "SRPMB"
  magnitude: number // Repurposed as SDS (Spectral Acceleration) multiplier (1 to 10 mapped to 0.1g to 1.5g)
}

export interface EarthquakeResult {
  damping: number
  stiffness: number
  materialLabel: string
  baseShear: number
  maxDrift: number
  allowableDrift: number
  /** Get amplitude at time t (seconds) */
  amplitudeAt: (t: number) => number
  /** Get floor deflections at time t */
  deflectionsAt: (t: number) => number[]
}

const MATERIAL_CONFIG = {
  SRPMB: { R: 3.0, Cd: 2.5, damping: 0.05, label: "Rangka Beton Biasa (SRPMB, R=3)" },
  SRPMM: { R: 5.0, Cd: 4.5, damping: 0.05, label: "Rangka Beton Menengah (SRPMM, R=5)" },
  SRPMK: { R: 8.0, Cd: 5.5, damping: 0.05, label: "Rangka Beton Khusus (SRPMK, R=8)" },
}

export function calculateEarthquake(params: EarthquakeParams): EarthquakeResult {
  const config = MATERIAL_CONFIG[params.material]

  // SNI 1726 Parameters
  const Sds = (params.magnitude / 10) * 1.5; // Scale 1-10 to 0.15g - 1.5g
  const Ie = 1.0; // Faktor Keutamaan
  const R = config.R;
  const Cd = config.Cd;
  
  // Approximate building properties
  const floorHeight = 4.0; // 4 meters per floor
  const hn = params.floors * floorHeight;
  const W = params.floors * 2500; // Assume 2500 kN weight per floor
  
  // Seismic Base Shear (V = Cs * W)
  let Cs = Sds / (R / Ie);
  if (Cs < 0.044 * Sds * Ie) Cs = 0.044 * Sds * Ie; // SNI minimum
  const V = Cs * W; // kN

  // Approximate Roof Drift (elastic) -> very simplified structural mechanics for cantilever
  // Delta_e = V * hn^3 / (3 * E * I_eq)
  // We'll calibrate an arbitrary stiffness so that drift is somewhat realistic (e.g. 1-5% of height)
  // Let's just use an empirical mapping to get Delta_elastic based on Cs
  const Delta_elastic = (Cs * hn) / 20.0; // rough empirical elastic drift in meters
  
  // Design Inelastic Drift (SNI 1726)
  const maxDrift = (Cd * Delta_elastic) / Ie; // meters
  
  // Allowable Drift (SNI 1726)
  const allowableDrift = 0.020 * hn;

  // For animation: damped harmonic oscillation, starting at 0, max reaching maxDrift
  // Period T = 0.0466 * hn^0.9
  const T = 0.0466 * Math.pow(hn, 0.9);
  const freq = 1 / T;
  
  const amplitudeAt = (t: number): number => {
    // scale to maxDrift, with damping envelope
    return maxDrift * Math.exp(-config.damping * 2 * t) * Math.sin(2 * Math.PI * freq * t);
  }

  const deflectionsAt = (t: number): number[] => {
    const amp = amplitudeAt(t);
    const floors = params.floors;
    const yCoords: number[] = [];
    for (let i = 0; i <= floors; i++) {
      const y = i / floors; // normalized 0..1
      const deflection = amp * Math.pow(y, 1.5); // 1.5 power for shear building mode shape
      yCoords.push(deflection);
    }
    return yCoords;
  }

  return {
    damping: config.damping,
    stiffness: 10, // arbitrary visual param
    materialLabel: config.label,
    baseShear: V,
    maxDrift,
    allowableDrift,
    amplitudeAt,
    deflectionsAt,
  }
}

/**
 * Get safety status based on SNI 1726 drift limits.
 */
export function getEarthquakeStatus(
  maxDrift: number,
  allowableDrift: number,
  material: string
): { status: "safe" | "warning" | "danger"; message: string } {
  
  if (maxDrift > allowableDrift * 1.5) {
    return {
      status: "danger",
      message: `BAHAYA RUNTUH (SNI 1726): Simpangan Atap (${maxDrift.toFixed(2)}m) jauh melebihi batas izin (${allowableDrift.toFixed(2)}m). Kolom beton berisiko patah geser! Gunakan sistem SRPMK atau Dinding Geser.`,
    }
  }
  if (maxDrift > allowableDrift) {
    return {
      status: "warning",
      message: `TIDAK MEMENUHI SYARAT: Simpangan Atap (${maxDrift.toFixed(2)}m) melewati batas izin SNI (${allowableDrift.toFixed(2)}m). Kerusakan non-struktural (dinding bata) akan sangat parah.`,
    }
  }
  return {
    status: "safe",
    message: `MEMENUHI SNI 1726: Simpangan Atap (${maxDrift.toFixed(2)}m) masih di bawah batas aman izin (${allowableDrift.toFixed(2)}m). Struktur dirancang daktail.`,
  }
}
